dids/src/main/kotlin/web5/sdk/dids/ion/model/Models.kt

```diff
@@ -0,0 +1,171 @@
+  val services: List<Service> = emptyList()
+)
+
+/**
+ * Represents an ION service.
+ *
+ * @property id The service ID.
+ * @property type The service type.
+ * @property serviceEndpoint The service endpoint.
+ */
+public data class Service(
+  public val id: String,
+  public val type: String,
+  public val serviceEndpoint: String
+)
+
+/**
+ * Represents a public key in the ION document as defined in item 3 of https://identity.foundation/sidetree/spec/#add-public-keys
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public data class PublicKey(
+  public val id: String,
+  public val type: String,
+  public val controller: String? = null,
+
+  @JsonSerialize(using = JacksonJWK.Serializer::class)
+  @JsonDeserialize(using = JacksonJWK.Deserializer::class)
+  public val publicKeyJwk: JWK? = null,
+  public val purposes: List<PublicKeyPurpose> = emptyList()
+)
+
+/**
+ * JacksonJWK is a utility class that facilitates serialization for [JWK] types, so that it's easy to integrate with any
+ * class that is meant to be serialized to/from JSON.
+ */
+public class JacksonJWK {
+  /**
+   * [Serializer] implements [JsonSerializer] for use with the [JsonSerialize] annotation from Jackson.
+   */
+  public object Serializer : JsonSerializer<JWK>() {
+    override fun serialize(value: JWK, gen: JsonGenerator, serializers: SerializerProvider) {
+      with(gen) {
+        writeObject(value.toJSONObject())
+      }
+    }
+  }
+
+  /**
+   * [Deserializer] implements [JsonDeserializer] for use with the [JsonDeserialize] annotation from Jackson.
+   */
+  public object Deserializer : JsonDeserializer<JWK>() {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext?): JWK {
+      @Suppress("UNCHECKED_CAST")
+      val node = p.readValueAs(Map::class.java) as MutableMap<String, Any>
+      return JWK.parse(node)
+    }
+  }
+}
+
+/**
+ * Enum representing the purpose of a public key.
+ */
+public enum class PublicKeyPurpose(@get:JsonValue public val code: String) {
+  AUTHENTICATION("authentication"),
+  KEY_AGREEMENT("keyAgreement"),
+  ASSERTION_METHOD("assertionMethod"),
+  CAPABILITY_DELEGATION("capabilityDelegation"),
+  CAPABILITY_INVOCATIO("capabilityInvocation"),
+}
+
+/**
+ * Sealed class representing a patch action in the ION document.
+ */
+@JsonTypeInfo(
+  use = JsonTypeInfo.Id.NAME,
+  include = JsonTypeInfo.As.PROPERTY,
+  property = "action"
+)
+@JsonSubTypes(
+  JsonSubTypes.Type(AddServicesAction::class, name = "add-services"),
+  JsonSubTypes.Type(ReplaceAction::class, name = "replace")
+)
+public sealed class PatchAction
+
+/**
+ * Represents an "add_services" patch action in the ION document.
+ *
+ * @property services List of services to add.
+ */
+public data class AddServicesAction(
+  public val services: List<Service> = emptyList()
+) : PatchAction()
+
+/**
+ * Represents a "replace" patch action in the ION document.
+ *
+ * @property document The document to replace.
+ */
+public data class ReplaceAction(
```

Comment:
would it be possible to include links to relevant sections of the sidetree spec for each of these data classes?